core/lib/zksync_core/src/api_server/tx_sender/master_pool_sink.rs

```diff
@@ -23,6 +30,23 @@ impl TxSink for MasterPoolSink {
         tx: L2Tx,
         execution_metrics: TransactionExecutionMetrics,
     ) -> Result<L2TxSubmissionResult, SubmitTxError> {
+        let address_and_nonce = (tx.initiator_account(), tx.nonce());
+
+        let mut lock = self.inflight_requests.lock().await;
+        match lock.entry(address_and_nonce) {
+            Entry::Occupied(entry) => {
+                let submission_res_handle = if entry.get() == &tx.hash() {
+                    L2TxSubmissionResult::Duplicate
+                } else {
+                    L2TxSubmissionResult::ReplacementInProgress
+                };
+                APP_METRICS.processed_txs[&TxStage::Mempool(submission_res_handle)].inc();
+                return Ok(submission_res_handle);
+            }
+            Entry::Vacant(entry) => entry.insert(tx.hash()),
```

Comment:
let's add a gauge for the size of the map. This might be useful to debug but even if not this is an insightful info to learn about the system